xmppserver/src/main/java/org/jivesoftware/util/SAXReaderUtil.java

```diff
@@ -0,0 +1,287 @@
+                                                                                    .setDynamic(false)
+                                                                                    .build();
+
+    public static final SystemProperty<Integer> PARSER_SERVICE_MAX_POOL_SIZE = SystemProperty.Builder.ofType(Integer.class)
+                                                                                    .setKey("xmpp.xmlutil.parser.maximum-pool-size")
+                                                                                    .setMinValue(1)
+                                                                                    .setDefaultValue(25)
+                                                                                    .setDynamic(false)
+                                                                                    .build();
+
+    public static final SystemProperty<Duration> PARSER_SERVICE_KEEP_ALIVE_TIME = SystemProperty.Builder.ofType(Duration.class)
+                                                                                    .setKey("xmpp.xmlutil.parser.keep_alive_time")
+                                                                                    .setChronoUnit(ChronoUnit.SECONDS)
+                                                                                    .setMinValue(Duration.ofMillis(0))
+                                                                                    .setDefaultValue(Duration.ofMinutes(1))
+                                                                                    .setDynamic(false)
+                                                                                    .build();
+
+    private static final ThreadPoolExecutor parserService = new ThreadPoolExecutor(
+                                                            PARSER_SERVICE_CORE_POOL_SIZE.getValue(),
+                                                            PARSER_SERVICE_MAX_POOL_SIZE.getValue(),
+                                                            PARSER_SERVICE_KEEP_ALIVE_TIME.getValue().toMillis(), TimeUnit.MILLISECONDS,
+                                                            new SynchronousQueue<>());
+
+    static
+    {
+        if (JMXManager.isEnabled()) {
+            final ThreadPoolExecutorDelegateMBean mBean = new ThreadPoolExecutorDelegate(parserService);
+            JMXManager.tryRegister(mBean, ThreadPoolExecutorDelegateMBean.BASE_OBJECT_NAME + "saxReaderUtil");
+        }
+    }
+
+    /**
+     * Parses an XML Document from the content provided by an input stream.
+     *
+     * The XML parsing is delegated to a limited set of parsers. When a parser is not immediately available, this method
+     * will block until one becomes available. The availability of parsers can be optimized by using the SystemProperties
+     * in this class.
+     *
+     * @param stream The data to be parsed as an XML Document.
+     * @return an XML Document.
```

Comment:
Might be worth adding an `@throws` expanding on why InterruptedException could be thrown (waiting for a thread) as that's not something you'd normally see parsing XML.

Reply:
(see also earlier comment on the ExecutionException)